### PR TITLE
Feature/improve app experience

### PR DESCRIPTION
**Pull request by liayoo**
- Now users can create "public" apps, which allow **anyone** to set values, rules, functions by default (no write_owner permission), in addition to the existing admin config.
  - Will be adding a util function for this feature in ain-js as well.
- Organized & added missing scripts in package.json.

### CODE REVIEW

File: db/functions.js [hunk 1]
```diff
@@ -543,31 +543,38 @@ class Functions {
...
-      const owner = {};
-      let rule = '';
-      const adminAddrList = Object.keys(adminConfig);
-      for (let i = 0; i < adminAddrList.length; i++) {
-        const addr = adminAddrList[i];
-        CommonUtil.setJsObject(owner, [OwnerProperties.OWNER, OwnerProperties.OWNERS, addr],
-            buildOwnerPermissions(true, true, true, true));
-        rule += `auth.addr === '${addr}'` + (i < adminAddrList.length - 1 ? ' || ' : '');
-      }
+    sanitizedVal[PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN] = adminConfig;
```
Comment (platfowner) [inline db/functions.js:557]: Could you consider extracting these sanitization parts into a separate function?

Comment (platfowner) [inline db/functions.js:557]: Could you consider rejecting the transactions with non-standard input for safety. We already have some reference code:
- https://github.com/ainblockchain/ain-blockchain/blob/develop/tx-pool/transaction.js#L239
- https://github.com/ainblockchain/ain-blockchain/blob/develop/db/index.js#L752

File: db/functions.js [hunk 2]
```diff
@@ -543,31 +543,38 @@ class Functions {
...
+    const appPath = PathUtil.getAppPath(appName);
+    const owner = {};
+    let rule = '';
+    const adminAddrList = Object.keys(adminConfig);
+    for (let i = 0; i < adminAddrList.length; i++) {
+      const addr = adminAddrList[i];
+      CommonUtil.setJsObject(owner, [OwnerProperties.OWNER, OwnerProperties.OWNERS, addr],
+          buildOwnerPermissions(true, true, true, true));
+      rule += `auth.addr === '${addr}'` + (i < adminAddrList.length - 1 ? ' || ' : '');
```
Comment (platfowner) [inline db/functions.js:566]: minor: you can consider using <array>.join() or _.join().
(ref: https://github.com/ainblockchain/ain-blockchain/blob/develop/db/state-node.js#L269)

File: db/functions.js [hunk 3]
```diff
@@ -543,31 +543,38 @@ class Functions {
...
-            buildOwnerPermissions(true, true, true, true));
-        rule += `auth.addr === '${addr}'` + (i < adminAddrList.length - 1 ? ' || ' : '');
-      }
+    sanitizedVal[PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN] = adminConfig;
+    const appPath = PathUtil.getAppPath(appName);
+    const owner = {};
+    let rule = '';
+    const adminAddrList = Object.keys(adminConfig);
+    for (let i = 0; i < adminAddrList.length; i++) {
+      const addr = adminAddrList[i];
```
Comment (platfowner) [inline db/functions.js:563]: applying CommonUtil.isCksumAddr()?
(ref: https://github.com/ainblockchain/ain-blockchain/blob/develop/db/state-util.js#L301)

File: db/functions.js [hunk 4]
```diff
@@ -543,31 +543,38 @@ class Functions {
     const recordId = context.params.record_id;
     const resultPath = PathUtil.getCreateAppResultPath(appName, recordId);
     const sanitizedVal = {};
-    const adminConfig = value[PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN];
+    const adminConfig = _.get(value, PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN);
     const billingConfig = _.get(value, PredefinedDbPaths.MANAGE_APP_CONFIG_BILLING);
     const serviceConfig = _.get(value, PredefinedDbPaths.MANAGE_APP_CONFIG_SERVICE);
+    const isPublic = _.get(value, PredefinedDbPaths.MANAGE_APP_CONFIG_IS_PUBLIC);
```
Comment (platfowner) [inline db/functions.js:549]: what about applying CommonUtil.isBool() to isPublic for sanity checking?

File: db/functions.js (hunk 1 above)
Comment (liayoo) [reply db/functions.js]: Ahhh that made _createApp much simpler..! Thanks for the suggestion!

Comment (liayoo) [reply db/functions.js]: Done!

File: db/functions.js (hunk 2 above)
Comment (liayoo) [reply db/functions.js]: Done!

File: db/functions.js (hunk 3 above)
Comment (liayoo) [reply db/functions.js:588]: The check is now included in `sanitizeCreateAppConfig()`. Thanks!!

File: db/functions.js (hunk 4 above)
Comment (liayoo) [reply db/functions.js]: Done, thanks!

File: db/functions.js [hunk 5]
```diff
@@ -536,44 +536,71 @@ class Functions {
...
+      return this.saveAndReturnFuncResult(context, resultPath, FunctionResultCode.INVALID_SERVICE_NAME);
+    }
+    const { sanitizedVal, errorCode } = this.sanitizeCreateAppConfig(value);
+    if (errorCode) {
+      return this.saveAndReturnFuncResult(context, resultPath, errorCode);
+    }
+    let rule;
+    const owner = {};
+    const adminAddrList = Object.keys(sanitizedVal[PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN]);
+    for (let i = 0; i < adminAddrList.length; i++) {
```
Comment (platfowner) [inline db/functions.js:587]: Do we need this for-loop?

Comment (platfowner) [reply db/functions.js:587]: Ah sorry, please ignore the previous comment.

File: db/functions.js [hunk 6]
```diff
@@ -536,44 +536,71 @@ class Functions {
...
+      const addr = adminAddrList[i];
+      rule = adminAddrList.map((addr) => `auth.addr === '${addr}'`).join(' || ');
```
Comment (platfowner) [inline db/functions.js:589]: can this line be moved to:
```
if (sanitizedVal[PredefinedDbPaths.MANAGE_APP_CONFIG_IS_PUBLIC]) {
  rule = true;
  ...
} else {
  // here?
}
```

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

File: db/functions.js (hunk 5 above)
Comment (liayoo) [reply db/functions.js]: Oops... thanks for catching it!

File: db/functions.js (hunk 6 above)
Comment (liayoo) [reply db/functions.js]: Done, thanks!

File: db/functions.js (hunk 5 above)
Comment (liayoo) [reply db/functions.js]: I was resetting rules over and over in the loop, so placed it outside (in the 'else' as you suggested) 🙂

**Comment by liayoo** [comment]
Thanks for the thorough review! Now merging~

**Pull request merged**